src/serialize.rs

```diff
@@ -1,67 +1,76 @@
 use std::io::Cursor;
 use std::io::Read;
-use std::io::Write;
+use std::io::{Error, ErrorKind};
 
 use crate::allocator::Allocator;
 use crate::node::Node;
 
 const MAX_SINGLE_BYTE: u8 = 0x7f;
 const CONS_BOX_MARKER: u8 = 0xff;
 
-fn encode_size(f: &mut dyn Write, size: usize) -> std::io::Result<()> {
+fn bad_encoding() -> std::io::Error {
+    Error::new(ErrorKind::InvalidInput, "bad encoding")
+}
+
+fn encode_size(out: &mut Vec<u8>, size: usize) {
     if size < 0x40 {
-        f.write_all(&[(0x80 | size) as u8])?;
+        out.push((0x80 | size) as u8);
     } else if size < 0x2000 {
-        f.write_all(&[(0xc0 | (size >> 8)) as u8, ((size) & 0xff) as u8])?;
+        out.extend(&[(0xc0 | (size >> 8)) as u8, ((size) & 0xff) as u8]);
```

Comment:
Is it `extend` for arrays, `push` for bytes?

Reply:
`extend()` is like `insert()` on std::vector